### ISSUE TITLE
Give the fall kill to player that hit the last

### ISSUE THREAD

**Issue by MetalGearDaner** [garbagemule/MobArena#415]
# Summary

* This issue is a…
    * [ ] Bug report
    * [X] Feature request
    * [ ] Documentation request
    * [ ] Other issue
    * [ ] Question
* **Describe the issue / feature in 1-2 sentences**: When mobs die by fall, fire, etc, give the kill to the last player that has hit the mob.

# Background

### Feature request 

* **Feature type**:
    * [ ] Arena configuration
    * [ ] Classes
    * [ ] Documentation
    * [ ] Economy integration
    * [ ] Leaderboards
    * [ ] New config settings
    * [X] Other (please explain)
    * [ ] Rewards
    * [ ] Third-party plugin support

**What does it do?**: So when you hit a mob (for example with knockback) and it dies because of falling, not the hit, it doesn't give you the score that you have killed that mob. This is specially bad for boss killing, because you may not get the reward if it dies not by your hits.

**Does it need new or changed commands? What are they?**: 

**Does this feature affect other parts of MobArena?**:

**Comment by ct-martin** [comment]
@garbagemule can you confirm whether MobArena has special handling for this, or whether it's using MC's detection

**Comment by garbagemule** [comment]
MobArena only checks for shooters/owners, so if the killer entity is a projectile or a pet, the shooter or owner will get the kill.

The logic sits in ArenaListener#onMonsterDeath.

**Comment by garbagemule** [comment]
Looks like the only way to really handle this is to keep track of the last player who "touched" the monster, but this is a whitelisting approach that is extremely difficult to get right. Consider plugins that give players magic spells where you can set monsters on fire - how is MobArena supposed to hook into _all_ such plugins and cover all bases?

I think we need to look into possible ways of dealing with this without having to create plugin dependencies everywhere. Thoughts, @NathanWolf ?

**Comment by NathanWolf** [comment]
Yeah this is a general issue for me with Quests and other plugins of that type.

I think the best thing you could do on your end is just track players that damage mobs. Unfortunately you can't use the last damage source for this, since that can change to fall, fire, poison, wither, etc damage (without a reference to the player).

But if you simply track the last player to damage, without clearing that if they receive non-player damage, I think that's good enough for my purposes.

I've been instructing people to always add a Damage action to spells that could potentially do damage (through fire, pushing, adding potion effects, etc), with a very tiny damage amount just so that it registers as having done damage.
